kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/dek/DataEncryptionKey.java

```diff
@@ -0,0 +1,333 @@
+            if (key != null) {
+                try {
+                    key.destroy();
+                }
+                catch (DestroyFailedException e) {
+                    var cls = key.getClass();
+                    if (LOGGER.isWarnEnabled()) {
+                        LOGGED_DESTROY_FAILED.computeIfAbsent(cls, (c) -> {
+                            LOGGER.warn("Failed to destroy an instance of {}. "
+                                    + "Note: this message is logged once per class even though there may be many occurrences of this event. "
+                                    + "This event can happen because the JRE's SecretKeySpec class does not override the destroy() method.",
+                                    c, e);
+                            return Boolean.TRUE;
+                        });
+                    }
+                }
+            }
+        }
+    }
+
+    public boolean isDestroyed() {
+        SecretKey secretKey = atomicKey.get();
+        return secretKey == null || secretKey.isDestroyed();
+    }
+
+    /**
+     * A means of performing a limited number of encryption operations without access to key material.
+     */
+    @NotThreadSafe
+    public final class Encryptor implements AutoCloseable {
+        private final Cipher cipher;
+        private SecretKey key;
+        private final Supplier<AlgorithmParameterSpec> paramSupplier;
+        private final CipherSpec cipherSpec;
+        private int numEncryptions;
+
+        private Encryptor(CipherSpec cipherSpec, SecretKey key, int numEncryptions) {
+            if (numEncryptions <= 0) {
+                throw new IllegalArgumentException();
+            }
+            this.cipher = cipherSpec.newCipher();
+            this.paramSupplier = cipherSpec.paramSupplier();
+            this.cipherSpec = cipherSpec;
+            this.key = key;
+            this.numEncryptions = numEncryptions;
+        }
+
+        /**
+         * @return The encrypted key
+         */
+        public E edek() {
+            return edek;
+        }
+
+        /**
+         * Perform an encryption operation using the DEK.
+         * @param plaintext The plaintext to be encrypted
+         * @param aad The AAD to be included in the encryption
+         * @param paramAllocator A function that will return a buffer into which the cipher parameters will be written.
+         * The function's argument is the number of bytes required for the cipher parameters.
+         * @param ciphertextAllocator A function that will return a buffer into which the ciphertext will be written.
+         * The function's first argument is the number of bytes required for the cipher parameters.
+         * The function's second argument is the number of bytes required for the ciphertext.
+         * @throws DekUsageException If this Encryptor has run out of operations.
+         */
+        public void encrypt(@NonNull ByteBuffer plaintext,
+                            @Nullable ByteBuffer aad,
```

Comment:
There's a bit of an assumption implicit in this signature: That the cipher is/supports AEAD. I don't think that's really a problem, because I don't think we'd be interested in ciphers/modes that weren't AEAD. But I could be wrong, and if I was then we'd need a way to know that a DEK didn't support AAD so we could complain up front about trying to use AAD with a cipher that didn't support it.